### PR TITLE
Show errors per input in `/eth-unit-conversion`

### PR DESCRIPTION
**Pull request by hasparus**
[DET-98](https://linear.app/deth/issue/DET-98/error-handling-per-component)

image

I had to remove `mx-auto`, because error messages resized the container when they appeared.
I also updated the error message for number length (the picture shows the old one). I'd rather use arbitrary precision numbers in the future though.

### CODE REVIEW

**Review by 7sne — APPROVED** [review]
this pr is bussin, respectfully

File: pages/eth-unit-conversion.page.tsx [hunk 1]
```diff
@@ -1,55 +1,70 @@
...
-type State = { wei: string; gwei: string; eth: string };
+type EthUnitConversionState = Record<
+  UnitType,
+  { value: string; error?: string }
+>;
+
+const powers = {
+  wei: -18,
+  gwei: -9,
+  eth: -1,
```
Comment (LogvinovLeon) [inline pages/eth-unit-conversion.page.tsx:19]: Not sure if we want to change that but `10^-1` is an unintuitive way to write `1`. Although coherent.
Screen Shot 2022-04-03 at 3 33 21 PM

File: pages/eth-unit-conversion.test.tsx [hunk 2]
```diff
@@ -7,16 +7,16 @@ describe(EthUnitConversion.name, () => {
...
 
     fireEvent.change(weiField, {
       target: { value: '10000000001000002333343' },
     });
 
     expect(weiField.value).toEqual('10000000001000002333343');
 
-    const gweiField = (await root.findByLabelText('gwei')) as HTMLInputElement;
-    const ethField = (await root.findByLabelText('eth')) as HTMLInputElement;
+    const gweiField = (await root.findByLabelText(/gwei/)) as HTMLInputElement;
```
Comment (LogvinovLeon) [inline pages/eth-unit-conversion.test.tsx:18]: Why do we want to use regexes where not necessary? To stay consistent with the assertions above? If so - shouldn't we change that in the whole project?

File: src/misc/constrain.ts [hunk 3]
```diff
@@ -0,0 +1,12 @@
+/**
+ * Ensures `value` is of a subtype of `Given` while preserving its exact type.
+ *
+ * @example
+ * const numbers = constrain<Record<string, number>>()({ one: 1, two: 2 })
+ *
+ * @see https://kentcdodds.com/blog/how-to-write-a-constrained-identity-function-in-typescript
+ */
+export const constrain =
```
Comment (LogvinovLeon) [inline src/misc/constrain.ts:9]: Noice ;)

**Review by LogvinovLeon — APPROVED** [review]

File: pages/eth-unit-conversion.test.tsx (hunk 2 above)
Comment (hasparus) [reply pages/eth-unit-conversion.test.tsx:18]: I moved the label a few levels up in the DOM so the hitbox for mouse clicks is more forgiving. The exact text inside the label might be something like "wei 10^-18" now. We also don’t care about casing here I think.

File: pages/eth-unit-conversion.page.tsx (hunk 1 above)
Comment (hasparus) [reply pages/eth-unit-conversion.page.tsx:19]: Yeah, I’ve been thinking about it. Was this in the designs? Shouldn’t we maybe just skip it?

**Pull request merged**